docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,208 @@
+
+To achieve this, ClusterResourceSet CRD is introduced that will be responsible for applying a set resources defined by users to the matching clusters.
+
+### Goals
+
+- Provide a means to specify a set of resources to apply automatically to newly-created and existing Clusters. Resources will be applied only once.
+- Support additions to the resource list by applying the new added resources to both new and existing matching clusters.
+- Support both json and yaml resources.
+
+### Non-Goals/Future Work
+
+- Replace or compete with the Cluster Addons subproject.
+- Support deletion of resources from clusters. Deleting a resource from a ClusterResourceSet or deleting a ClusterResourceSet does not result in deletion of those resources from clusters.
+- Lifecycle management of the installed resources (such as CNI).
+- Support reconciliation of resources on resource hash change and/or periodically. This can be a future enhancement work.
+
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As someone creating multiple clusters, I want all my clusters to have a CNI provider of my choosing installed automatically, so I don’t have to manually repeat the installation for each new cluster.
+
+#### Story 2
+
+As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.
+
+#### Story 3
+
+As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.
+
+### Implementation Details/Notes/Constraints
+
+#### Data model changes
+
+None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters (in any namespace).
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce"
+ clusterSelector:
+   matchLabels:
+     postcreatelabelcni: calico
+ resources:
```

Comment:
Does cluster API plan to support different sources like helm, git etc to fetch the resources from? If so, then it might be valuable to consider kapp-controller's approach to fetch resources (https://github.com/k14s/kapp-controller/blob/master/pkg/apis/kappctrl/v1alpha1/types_fetch.go#L7)